core/src/main/java/feign/ObjectParamMetadata.java

```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright 2012-2018 The Feign Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package feign;
+
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+class ObjectParamMetadata {
+
+  private final static ConcurrentMap<Class<?>, ObjectParamMetadata> classToMetadata =
+      new ConcurrentHashMap<Class<?>, ObjectParamMetadata>();
```

Comment:
I'm not sure if this cache is a case of premature optimization or not.

Reply:
...or if it belongs in `MethodMetadata` instead.

Reply:
I think I start getting the point of this class. Is it basically a `QueryObjectParamMapper` or so? as in - it's job is to, for a single object type, generate a `Map<String,Object>`?

If so, I think I get where you're headed and would suggest the following changes:
* Rename it to something closer to the description above. I like `QueryObjectParamMapper`, but if you think of a better one, thumbs up.
* The caching is likely not a premature optimisation (based on how many other places we try to figure out a class and then cache its metadata to not have it done again) - but is likely in the wrong place (why would it be done in this object?). Perhaps moving it to the place where it gets called from (`ReflectiveFeign`?) is a more appropriate fit? It also doesn't have to be `static` there - which is usually a bit of a code-smell.
* likely rename `toQueryMap` to something more like `mapObject(object)` or `convertToQueryMap` - the mapper isn't becoming a query map itself - it's converting an object into one - so make it that descriptive
* likely rename `getMetadata` to `generateMapper` (if you end up calling this a mapper) - and you can likely then make it the single-entry point on startup that I mentioned.

Thoughts?

Reply:
oh! and given the somewhat more complex nature - add some unit tests specifically to it - to make sure you capture intended behaviour and can replicate error cases easier. (test both correct behaviour and expected error cases)

Reply:
I read some more on your code. I agree on the right place being next to `indexToExpander`. If you want I can fork off your repo, give it a try, and then create PR into your repo for those changes so you can see what I mean/accept if you like.

Reply:
I named it based on the other metadata class, because that's what it *stores*. But you are right, what it *does* is turn an object into a map, so I'll rename it.

Some possibilities: `QueryObjectParamMapper`, `ObjectQueryParamMapper`, `QueryMapBuilder`, or keep the cache internal and call it `QueryMapFactory` (either keep it static, or make it an interface default implementation that you can add as a `Feign.Builder` option and pass it into the `ReflectiveFeign` constructor)? It actually is building a `Map` of query params (so it is a `QueryMapBuilder` or "factory").

How does that sound?

Reply:
@saintf I don't mind working on this. It's a good learning experience for me. But if you want to write some code on it, I'm happy to do look at that too.